GH-95704: Don't suppress errors from tasks when TG is cancelled

gvanrossum opened the pull request:
When a task catches CancelledError and raises some other error,
the other error should not silently be suppressed.

Exception: when uncancel() != 0, always propagate CancelledError.

NOTE: This represents a change in behavior (hence the need to
change three tests).  But it is only an edge case.

We need to discuss with the RM whether to backport to 3.11RC2,
or 3.11.1, or not at all.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-95704 -->
* Issue: gh-95704
<!-- /gh-issue-number -->

gvanrossum committed:
GH-95704: Don't suppress errors from tasks when TG is cancelled

When a task catches CancelledError and raises some other error,
the other error should not silently be suppressed.

Exception: when uncancel() != 0, always propagate CancelledError.

NOTE: This represents a change in behavior (hence the need to
change three tests).  But it is only an edge case.

We need to discuss with the RM whether to backport to 3.11RC2,
or 3.11.1, or not at all.

Tinche reviewed (changes requested):
In the spirit of the old boy scouts rule, 'Always leave the campground cleaner than you found it', can we give the changed tests meaningful names and docstrings?

The functionality change looks ok, although I was a little confused at first as to why we're now also checking `self._parent_cancel_requested`, so might want to add a small comment there?

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

gvanrossum committed:
Improve comment explaining when/why we propagate cancellations

graingert commented on Lib/asyncio/taskgroups.py:
```diff
@@ -115,10 +115,11 @@ async def __aexit__(self, et, exc, tb):
+        # Propagate CancelledError if there is one, except if there
+        # are other errors and we're not required by uncancel() != 0.
+        # (The latter means we *must* propagate the cancellation.)
+        if (propagate_cancellation_error is not None and
```
Doesn't this now suppress the TimeoutError when used with `async with timeout():`?

I'll try and make a repro

gvanrossum replied:
Since the condition is true in fewer cases than the original code (where it was just `if propagate_cancellation_error is not None:`) I don't see how it could suppress something that wouldn't be suppressed previously. But I'm not sure what you're thinking of, so please make a repro.

graingert commented on Misc/NEWS.d/next/Library/2022-08-08-01-42-11.gh-issue-95704.MOPFfX.rst:
```diff
@@ -0,0 +1,2 @@
+When a task catches CancelledError and raises some other error,
```
```suggestion
When a task catches :exc:`asyncio.CancelledError` and raises some other error,

```

gvanrossum committed:
Add markup to NEWS file

Co-authored-by: Thomas Grainger <tagrain@gmail.com>

gvanrossum committed:
Never ignore side errors

Any scenario where a task crashes in cleanup upon cancellation
will now result in an ExceptionGroup wrapping the crash(es)
instead of propagating CancelledError and ignoring the side errors.

Tinche reviewed (approved):
This LGTM. 

I was warming up to the cleverness of Yury's idea but if you think there be dragons I'm convinced.

gvanrossum committed:
Remove blank line from middle of blurb
